typhos/notes.py

```diff
@@ -0,0 +1,189 @@
+import os
+from datetime import datetime
+from enum import IntEnum
+from pathlib import Path
+from typing import Dict, Optional, Tuple
+
+import platformdirs
+import yaml
+from qtpy import QtWidgets
+
+from typhos import utils
+
+logger = logging.getLogger(__name__)
+NOTES_VAR = "PCDS_DEVICE_NOTES"
+
+
+class NotesSource(IntEnum):
+    USER = 0
+    ENV = 1
+    # HAPPI = 2  # to be implemented later
+
+
+def get_data_from_yaml(device_name: str, path: Path) -> Optional[Dict[str, str]]:
+    """
+    Returns the device data from the yaml file stored in ``path``.
+    Returns `None` if reading the file fails or there is no information
+
+    Parameters
+    ----------
+    device_name : str
+        The name of the device to retrieve notes for.
+    path : Path
+        Path to the device information yaml
+
+    Returns
+    -------
+    Optional[Dict[str, str]]
+        The device information or None
+    """
+    try:
+        with open(path) as f:
+            device_notes = yaml.full_load(f)
+    except Exception as ex:
+        logger.warning(f'failed to load device notes: {ex}')
+        return
+
+    if device_name in device_notes:
+        return device_notes.get(device_name, None)
+
+
+def get_notes_data(device_name: str) -> Tuple[NotesSource, Dict[str, str]]:
+    """
+    get the notes data for a given device
+    attempt to get the info from the following locations
+    in order of priority (higher priority will shadow)
+    1: device_notes in the user directory
+    2: the path in NOTES_VAR environment variable
+    3: TODO: happi ... eventually
+
+    Parameters
+    ----------
+    device_name : str
+        The device name.  Can also be a component. e.g. device_component_name
+
+    Returns
+    -------
+    Tuple[NotesSource, dict]
+        The source of the device notes, and
+        a dictionary containing the device note information
+    """
+    data = {'note': '', 'timestamp': ''}
+    source = NotesSource.USER
+
+    # try env var
+    env_notes_path = os.environ.get(NOTES_VAR)
+    if env_notes_path and Path(env_notes_path).is_file():
+        note_data = get_data_from_yaml(device_name, env_notes_path)
+        if note_data:
+            data = note_data
+            source = NotesSource.ENV
+
+    # try user directory
+    user_data_path = platformdirs.user_data_path() / 'device_notes.yaml'
+    if user_data_path.exists():
+        note_data = get_data_from_yaml(device_name, user_data_path)
+        if note_data:
+            data = note_data
+            source = NotesSource.USER
+
+    return source, data
+
+
+def insert_into_yaml(path: Path, device_name: str, data: dict[str, str]) -> None:
+    try:
+        with open(path, 'r') as f:
+            device_notes = yaml.full_load(f)
+    except Exception as ex:
+        logger.warning(f'unable to open existing device info: {ex}')
+        return
+
+    device_notes[device_name] = data
+
+    try:
+        with open(path, 'w') as f:
+            yaml.dump(device_notes, f)
```

Comment:
This should use the "make a second file and copy over the first" method to make sure we do an atomic write

Reply:
ah right I forgot about that